# Methods
### Choosing sequences:
Ten genomic sequences from 10 different patients were downloaded from the NCBI database as
CSV file format. Complete RNA coronavirus genome, the region was selected “Egypt”, and
the date was the newest. This was the criteria we followed.
4.2 Selecting genes:
“S”, “E”, “N”, and “RDRP” genes were selected from genomic sequences, as from each
genomic sequence we downloaded the four genes in Fasta file format.
### Alignment:
MSA was done for each gene using Clustal Omega; first alignment for the “S” gene, second alignment for the “E” gene, third alignment for the “N” gene, and
fourth alignment for the “RDRP” gene.
### Visualization:
Visualization of multiple gene alignment was done using Jalview software.
### Conservation estimation:
Conservation estimation was done by using bio-python.
